config/config_mgmt.py

```diff
@@ -0,0 +1,864 @@
+            for waitTime in range(timeout):
+                self.sysLog(msg='Check Asic DB: {} try'.format(waitTime+1))
+                # checkNoPortsInAsicDb will return True if all ports are not
+                # present in ASIC DB
+                if self._checkNoPortsInAsicDb(db, ports, portMap):
+                    break
+                tsleep(1)
+
+            # raise if timer expired
+            if waitTime + 1 == timeout:
+                print("!!!  Critical Failure, Ports are not Deleted from \
+                    ASIC DB, Bail Out  !!!")
+                self.sysLog(syslog.LOG_CRIT, "!!!  Critical Failure, Ports \
+                    are not Deleted from ASIC DB, Bail Out  !!!")
+                raise(Exception("Ports are present in ASIC DB after timeout"))
+
+        except Exception as e:
+            print(e)
+            raise e
+
+        return True
+
+    def breakOutPort(self, delPorts=list(), addPorts= list(), portJson=dict(), \
+        force=False, loadDefConfig=True):
+        '''
+        This is the main function for port breakout. Exposed to caller.
+
+        Parameters:
+            delPorts (list): ports to be deleted.
+            addPorts (list): ports to be added.
+            portJson (dict): Config DB json Part of all Ports, generated from
+                platform.json.
+            force (bool): if false return dependecies, else delete dependencies.
+            loadDefConfig: If loadDefConfig, add default config for ports as well.
+
+        Returns:
+            (deps, ret) (tuple)[list, bool]: dependecies and success/failure.
+        '''
+        MAX_WAIT = 60
+        try:
+            # delete Port and get the Config diff, deps and True/False
+            delConfigToLoad, deps, ret = self._deletePorts(ports=delPorts, \
+                force=force)
+            # return dependencies if delete port fails
+            if ret == False:
+                return deps, ret
+
+            # add Ports and get the config diff and True/False
+            addConfigtoLoad, ret = self._addPorts(ports=addPorts, \
+                portJson=portJson, loadDefConfig=loadDefConfig)
+            # return if ret is False, Great thing, no change is done in Config
+            if ret == False:
+                return None, ret
+
+            # Save Port OIDs Mapping Before Deleting Port
+            dataBase = SonicV2Connector(host="127.0.0.1")
+            if_name_map, if_oid_map = port_util.get_interface_oid_map(dataBase)
+            self.sysLog(syslog.LOG_DEBUG, 'if_name_map {}'.format(if_name_map))
+
+            # If we are here, then get ready to update the Config DB, Update
+            # deletion of Config first, then verify in Asic DB for port deletion,
+            # then update addition of ports in config DB.
+            self.writeConfigDB(delConfigToLoad)
+            # Verify in Asic DB,
+            self._verifyAsicDB(db=dataBase, ports=delPorts, portMap=if_name_map, \
+                timeout=MAX_WAIT)
```

Comment:
@renukamanavalan , I assumes, you are asking if writing delConfigToLoad to config DB, results in crash of orchgent & syncd, then system is in inconsistent state. Also db is not correct because that only caused the crash. Which is the risk as today, where pushing a wrong config to DB may result in crash.

To recover from it, we are working on creating rollback config and push it to DB. But it will not be straight forward because we need to warm-reboot SWSS and syncd to recover from it. So that part will take some effort to implement.

Reply:
Not even that, what if syncd & orchagent were already down, for whatever reason ... explicitly stopped/crashed due to some other issue.

The point is that the port breakdown, is not atomic, but pause in the middle for orchagent & syncd to do their job and then proceed.

If this dependency can't be avoided, can we have a diff to rollback to, for any exception failure, so we leave the DB at whatever the state it was in. This is ideal

Reply:
This seems same as running 'config load' command when orchagent is down. So no new risk is present right?

Reply:
I don't think config load command relies on the chain reaction of CONFIG --> APP --> ASIC. Do you know any ?

Reply:
Log a new issue for this rollback config.

Reply:
https://github.com/Azure/sonic-utilities/issues/964.